src/storage/buxfer.ts

```diff
@@ -0,0 +1,121 @@
+import {
+  BUXFER_USER_NAME,
+  BUXFER_PASSWORD,
+  BUXFER_ACCOUNTS,
+} from "../config.js";
+import { SaveStats, TransactionRow, TransactionStorage } from "../types.js";
+import { createLogger } from "./../utils/logger.js";
+import { parseISO, format } from "date-fns";
+import { TransactionStatuses } from "israeli-bank-scrapers/lib/transactions.js";
+import {
+  BuxferApiClient,
+  BuxferTransaction,
+  AddTransactionsResponse,
+} from "buxfer-ts-client";
+
+const BUXFER_DATE_FORMAT = "yyyy-MM-dd";
+const logger = createLogger("BuxferStorage");
+
+export class BuxferStorage implements TransactionStorage {
+  private buxferClient: BuxferApiClient;
+  private accountToBuxferAccount: Map<string, string>;
+
+  async init() {
+    logger("init");
+    this.buxferClient = new BuxferApiClient(BUXFER_USER_NAME, BUXFER_PASSWORD);
+    this.accountToBuxferAccount = this.parseBuxferAccounts(BUXFER_ACCOUNTS);
+  }
+
+  canSave() {
+    return Boolean(BUXFER_USER_NAME && BUXFER_PASSWORD && BUXFER_ACCOUNTS);
+  }
+
+  async saveTransactions(txns: Array<TransactionRow>) {
+    await this.init();
+
+    const stats = {
+      name: "BuxferStorage",
+      table: `Accounts: "${Array.from(this.accountToBuxferAccount.keys())}"`,
+      total: txns.length,
+      added: 0,
+      pending: 0,
+      existing: 0,
+      skipped: 0,
+    } satisfies SaveStats;
+
+    // Initialize an array to store non-pending and non-empty account ID transactions on Buxfer format.
+    const txToSend: BuxferTransaction[] = [];
+    const missingAccounts = new Set<string>();
+
+    for (const tx of txns) {
+      const accountId = this.accountToBuxferAccount.get(tx.account);
+      if (!accountId) {
+        missingAccounts.add(tx.account);
+        stats.skipped++;
+        continue;
+      }
+
+      // Converting to Buxfer format.
+      const buxferTx = this.convertTransactionToBuxferFormat(tx, accountId);
+
+      // Add non-pending and non-empty account ID transactions to the array.
+      txToSend.push(buxferTx);
+    }
+
+    if (txToSend.length > 0) {
+      // TODO - Build JSON based personal rule engine for tagging transactions
+      this.tagTransactionsByRules(txToSend);
+
+      // Send transactions to Buxfer
+      logger(
+        `sending to Buxfer accounts: "${this.accountToBuxferAccount.keys()}"`,
+      );
+      const resp: AddTransactionsResponse =
+        await this.buxferClient.addTransactions(txToSend, true);
+      logger("transactions sent to Buxfer successfully!");
+      stats.added = resp.addedTransactionIds.length;
+      stats.existing = resp.duplicatedTransactionIds.length;
+      stats.skipped += stats.existing;
+    }
+
+    if (missingAccounts.size > 0) {
+      logger(`Accounts missing in Buxfer_ACCOUNTS:`, missingAccounts);
+    }
+
+    return stats;
+  }
+
+  tagTransactionsByRules(txToSend: BuxferTransaction[]) {
+    // TODO - Implement declarative rule engine
```

Comment:
Do you have any idea on how you want to create it?
See #120

Reply:
Well I have not tested or attempted a POC yet, but generally I think its best to have a set of JSON file rules that each user could define for himself **and not committed.** The rules would define the auto tagging logic for each transaction to be uploaded to storages.
I am having a look at some packages that can accomplish this:
https://github.com/CacheControl/json-rules-engine
https://github.com/andrewvo89/rules-engine-ts